design/vouching-contracts-v2.1.md

```diff
@@ -0,0 +1,178 @@
+  requires amount > 0
+  requires amount <= vouched(msg.sender, id)
+  requires vouched(msg.sender, id) - amount >= minStake if msg.sender == owner(id)
+  emits Unvouch(uint256 indexed id, address indexed sender, uint256 amount)
+```
+
+Decreases the vouch for the package identified by `id` by `amount` for `sender`. Note that if `sender` is the `vouched` owner, he cannot decrease his vouching under `minStake`.
+
+### Move
+
+```
+move(uint256 from, uint256 to, uint256 amount)
+  requires amount > 0
+  requires amount <= vouched(msg.sender, from)
+  requires packages from and to to exist
+  requires msg.sender != owner(from) || msg.sender == owner(to) || (vouched(msg.sender, id) - amount >= minStake)
+  emits Vouch(uint256 indexed id = from, address indexed sender, uint256 amount)
+  emits Unvouch(uint256 indexed id = to, address indexed sender, uint256 amount)
+```
+
+Moves vouching from a package to another by `amount`. Note that if the owner is moving stake to another package he owns as well, then the restriction of keeping at least minStake does not apply. This enables a maintainer to remove all vouching from a version to move it to a different one, without any penalties. It also enables any user to shift their vouching without any cost.
+
+### Transfer ownership
+
+```
+transfer(uint256 id, address newOwner, uint256 amount)
+  requires owner(id) == msg.sender
+  requires vouched(id, msg.sender) >= amount
+  requires vouched(id, newOwner) + amount >= minStake
+  emit Transfer(address from = msg.sender, address to = newOwner, uint256 id)
+```
+
+Transfers ownership of a package to another address, optionally transferring `amount` of the tokens vouched for this package to the new owner as well. The new owner must end up with at least `minStake` tokens vouched for the package (note that the new owner may had already vouched for the package, and these tokens count towards the minStake requirement). 
+
+_We may decide to not implement this function at this stage._
+
+### Set metadata URI
+
+```
+setMetadataURI(uint256 id, string metadataURI)
+  requires owner(id) == msg.sender
+  emits MetadataURIChanged(uint256 indexed id, string metadataURI)
+```
+
+Changes the URI where the metadata is stored. Note that it is not possible to change the metadata hash, so only the location can be changed, not the contents.
+
+_We may decide to not implement this function at this stage._
+
+### Vouching getters
+
+```
+vouched(id) returns address
+metadata(id) returns (string uri, bytes32 hash)
+owner(id) returns address
+totalVouched(id) returns uint256
+vouched(id, address) returns uint256
+```
+
+Returns information on a `vouched` item.
+
+## Challenges
+
+The vouching contract should expose the following interface for managing challenges to vouched items. These methods will be implemented in a second stage, and are subject to change.
+
+### Challenge
+
+```
+challenge(uint256 id, uint256 amount, string challengeURI, bytes32 challengeHash)
```

Comment:
The `id` (or address) of the targeted entity is in `challengeHash`? Shouldn't we pass it as a parameter here?

Reply:
Oh, no, sorry, I was confused because of the comment below in L125: 

> Creates a new challenge with a fresh id in a _pending_ state...

I thought `uint256 id` here was the `challengeId` (which of course wouldn't make sense), and I had the package `id` missing :)

I'd maybe modify said comment to read:

> Creates a new challenge with a fresh challengeID in a _pending_ state...